pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -263,6 +264,18 @@ public class Utils {
         return nodes
     }
 
+    static void markExecutedStagesOnAction(CpsScript script, int astHashCode) throws Exception {
+        WorkflowRun r = script.$build()
+        ExecutionModelAction action = r.getAction(ExecutionModelAction.class)
+        if (action != null) {
+            if (action.stagesHashCode > 0) {
+                throw new IllegalStateException("Only one pipeline { ... } block can be executed in a single run.")
+            }
+            action.setStagesHashCode(astHashCode)
+            r.save()
```

Comment:
Do you have to save here? Pipelines are hitting the master's storage too much as it is.

Reply:
Yeah, we do - the action already exists, so just modifying it won't result in it being saved. Also, this literally happens only once per build (since if this method gets called a second time, we throw that exception above!)

Reply:
Also, this is writing the run's `config.xml`, not the XML storage for an individual `FlowNode`, so it's no worse than, say, setting the build result, etc.

Reply:
Well my other concern was around badly coded `SaveableListener`s out there that might get confused with more than one save per build.
I would guess the build is saved at the end as for freestyles but there might be more to it here if it should survive restarts.

Reply:
Oh, no, `run.save()` gets called a *bunch* already - like I said, any time the build result gets set, a new action is added, you name it.